python/erdos/timestamp.py

```diff
@@ -1,37 +1,81 @@
 from typing import Sequence
+from erdos.internal import PyTimestamp
 
 
 class Timestamp(object):
-    """An ERDOS timestamp.
-   The timestamp can consist of one or more coordinates.
+    """An ERDOS timestamp representing the time for which the messages or
+   watermarks are sent.
+
+   A :py:class:`Timestamp` can have one of the following three values:
+    - Timestamp::Bottom: Representing the initial time, and used for 
+        initializing operators.
+    - Timestamp::Time(c): An integer timestamp representing the time attached 
+        to the message or watermark.
+    - Timestamp::Top: The final timestamp, and used for destruction of 
+        operators.
 
    Attributes:
-       timestamp (:py:class:`.Timestamp`): For the copy constructor.
        coordinates (`Sequence[int]`): An array of coordinates.
-       is_top (`bool`): Whether this is the highest possible timestamp.
+       is_bottom (`bool`): Representing Timestamp::Bottom.
+       is_top (`bool`): Representing Timestamp::Top.
     """
     def __init__(self,
                  timestamp=None,
                  coordinates: Sequence[int] = None,
-                 is_top: bool = False):
-        self._is_top = is_top
-        if is_top and coordinates is None:
-            coordinates = []
-        if timestamp is None:
-            assert coordinates is not None, "Timestamp has empty coordinates"
-            self.coordinates = coordinates
-        else:
+                 is_top: bool = False,
+                 is_bottom: bool = False,
+                 _py_timestamp: PyTimestamp = None):
+        if _py_timestamp is not None:
+            # Initialize from PyTimestamp, if available.
+            self._is_top = _py_timestamp.is_top
+            self._is_bottom = _py_timestamp.is_bottom
+            self.coordinates = _py_timestamp.coordinates
+        elif timestamp is not None:
+            # If Timestamp is available, copy its contents.
+            self._is_top = timestamp.is_top
+            self._is_bottom = timestamp.is_bottom
             self.coordinates = timestamp.coordinates
+        else:
+            if is_top and not is_bottom and coordinates is None:
+                self._is_top = is_top
+                self._is_bottom = is_bottom
+                self.coordinates = coordinates
+            elif is_bottom and not is_top and coordinates is None:
+                self._is_top = is_top
+                self._is_bottom = is_bottom
+                self.coordinates = coordinates
+            elif coordinates is not None and not is_bottom and not is_top:
+                self._is_top = is_top
+                self._is_bottom = is_bottom
+                self.coordinates = coordinates
+            else:
+                raise ValueError("Timestamp should either have coordinates"
+                                 "or be either Top or Bottom")
```

Comment:
Might be easier to maintain if we always instantiate a `PyTimestamp`, and store `self._py_timestamp`.

Then we can re-use the Rust implementation for `__lt__`, `__leq__` by writing it in Rust in the `PyTimestamp` class. Properties like `is_top` would return `self._py_timestamp.is_top`, etc.

Reply:
We might also want to consider moving the `Timestamp` class to pure Rust and getting rid of this Python wrapper. I believe pyo3 allows us to add docstrings from Rust now too.

Thoughts?

Reply:
I wrote a minor PyTimestamp wrapper that patches almost all methods from Python to Rust so we utilize our implementations in Rust now for comparisons.